### PR TITLE
refactor(core/presentation): Refactor form inputs and layouts for better composability and reuse.

### PR DESCRIPTION
**Pull request by christopherthielen**
This is an iteration of the Spinnaker Forms API.  If I pinged you on this PR you expressed interest in this update.  Review if you want.

- Introduce `<FormikForm render={formikProps => { render the form values }} />` component
- Introduce `<FormField/>` and `<FormikFormField/` components with render props
- Remove `formikField` HOC in favor of `<FormikFormField/>`
- Remove `TextField` in favor of using `<FormikFormField input={TextInput} />`
- Rename `BasicLayout` to `StandardFieldLayout`
- Refactor Expression Form Fields to use new API/Components

# React Form Components

# FormikFormField

`FormikFormField` is a component used to render form fields in Spinnaker
using Formik for form state management and data validation.

basic formik form

This example sets up a Formik form and renders two `FormikFormField`s.
Validation errors on each field are surfaced when the field is blurred.
It has a submit button which is disabled when the form is invalid or being submitted.

```js
<Formik
  initialValues={{ name: '', email: '' }}
  onSubmit={(_values, formik) => setTimeout(() => formik.setSubmitting(false), 2000)}
  render={formik => {
    return (
      <Form>
        <FormikFormField
          name="name"
          label="Your Name"
          input={TextInput}
          validate={value => (!value ? 'Please enter your name' : undefined)}
        />

        <FormikFormField
          name="email"
          label="Your Email"
          input={TextInput}
          validate={value => {
            if (!value) return 'Please enter your email';
            if (!/[^@]+@[^@]+/.exec(value)) return 'Please enter a valid email';
          }}
        />

        <button disabled={!formik.isValid || formik.isSubmitting} className={`pull-right primary`}>
          {formik.isSubmitting ? 'Submitting...' : 'Submit Form'}
        </button>
      </Form>
    );
  }}
/>
```

The `FormikFormField` component should be a descendent of a `Formik` component.
It accepts "all the props", organizes them, and passes them down to the Input and Layout in the correct spots.

- `input`: the Input component to use (see Input section below for details)
- `layout`: the (optional) Layout component to use. (see Layout section for details. `StandardFieldLayout` is used by default)
- `name`: the path to the field's value in the formik `values`
- `label`, `help`, `required`, `actions` (see `IFieldLayoutPropsWithoutInput`)
- `touched`, `validationMessage`, `validationStatus` (see: `IValidationProps`)

A field validation function can be attached directly to the `FormikFormField` component via the `validate` prop,
or the field can be validated at the form level using the `Formik` component's `validate` prop.

### Form level validation Example:

This example shows validation of a formik field using the `Formik` component's `validate` prop.

```js
<Formik
  initialValues={{ email: '' }}
  validate={values => {
    const errors = {};
    if (!value) {
      errors.email = 'Please enter your email';
    } else if (!/[^@]+@[^@]+/.exec(value)) {
      errors.email = 'Please enter a valid email';
    }
    return errors;
  }}
  render={formik => {
    return (
      <Form>
        <FormikFormField name="email" label="Your Email" input={TextInput} />
      </Form>
    );
  }}
/>
```

# FormField

`FormField` is a component used to render form fields in Spinnaker
using externally managed form state and validation.

### TL;DR Example:

This example renders two `FormField`s and does field level validation.
This trivial example does not try to manage form-level validation state.
It has a submit button which is disabled when the form is being submitted.

```js
<form
  onSubmit={() => {
    this.setState({ isSubmitting: true });
    const stopSubmitting = () => this.setState({ isSubmitting: false });
    this.submit().then(stopSubmitting, stopSubmitting);
  }}
>
  <FormField
    name="name"
    label="your name"
    input={TextInput}
    value={this.state.name}
    onChange={evt => this.setState({ name: evt.target.value })}
    validate={val => !val && <span>Please enter your name</span>}
  />
  <FormField
    name="email"
    label="email address"
    input={TextInput}
    value={this.state.email}
    onChange={evt => this.setState({ email: evt.target.value })}
    validate={val => val && val.indexOf('@') === -1 && <span>Please enter a valid email</span>}
  />
  <button disabled={this.state.isSubmitting} className={`pull-right primary`}>
    {this.state.isSubmitting ? 'Submitting...' : 'Submit Form'}
  </button>
</form>
```

The `FormField` component accepts "all the props", organizes them,
and passes them down to the Input and Layout in the correct spots.

The props accepted by FormField are:

- `input`: the Input component to use
- `layout`: the Layout component to use. (optional, `StandardFieldLayout` is used by default)
- `label`, `help`, `required`, `actions` (see `IFieldLayoutPropsWithoutInput`)
- `name`, `value`, `onChange`, `onBlur`, (see `IControlledInputProps`)
- `touched`, `validationMessage`, `validationStatus` (see: `IValidationProps`)

### Custom Input Example:

This example uses an input which is defined inline:

```js
<FormField
  name="email"
  label="email address"
  input={({ field, validation }) => <input type="text" className={!!validation.error && 'error'} {...field} />}
  value={this.state.email}
  onChange={evt => this.setState({ email: evt.target.value })}
  validate={val => val && val.indexOf('@') === -1 && <span>Please enter a valid email</span>}
/>
```

### Layout Example:

This example uses a custom layout to render the error above the input

```js
<FormField
  name="email"
  label="email address"
  input={TextInput}
  layout={({ error, input }) => (
    <div>
      <div style={{ background: 'red' }}>{error}</div>
      <div>{input}</div>
    </div>
  )}
  value={this.state.email}
  onChange={evt => this.setState({ email: evt.target.value })}
  validate={val => val && val.indexOf('@') === -1 && <span>Please enter a valid email</span>}
/>
```

# Inputs

An Input is a react component which is rendered by a `FormField` or `FormikFormField` and:

- accepts user input
- emits change events when the input changes
- visually indicates an invalid validation state

An Input receives two important props (see `IFormInputProps`):

- `field`: an `IControlledInputProps` object which provides convenient props for rendering controlled form inputs.
  - `name`, `value`, `onChange`, `onBlur`
  - The contents of this prop is typically spread onto an input, e.g., `<input {...field} />`
- `validation`: Can be used to visually indicate the current validation state of the input
  - `validationMessage`: The validation message -- in general, render the message in a Layout, not here
  - `validationStatus`: The validation status -- visually indicate the validation status (See IValidationStatus)
  - `touched`: a flag which indicates if the the input has been "touched", which generally means "blurred"

### Example:

```
const SimpleTextInput = ({ field, validation }: IFormInputProps) =>
  <input type="text" className={validation.validationStatus} {...field} />
```

An Input can accept additional props, as needed.

### Example:

```
const SimpleTextInput = ({ field, validation, text }: IFormInputProps & { text: string }) =>
  <span> <input type="checkbox" {...field} /> {text} </span>
```

Commonly used inputs such as `type="text"` should be should be encapsulated in a reusable component, such as `TextInput`.

# Layouts

A Layout component accepts the various elements of full form field.
It defines how these components are visually rendered in relation to each other.
The Layout is also responsible for implementing responsive design, and can adjust element positions based on screen size.

The components that a Layout manages are:

- Input
- Label
- Help widget (i.e., popover)
- Required field indicator
- Action Icons (trash, etc)
- Validation message

### Example Layout:

```
+-----------------------------------------------------------------+
+ * Label Text [Help] | Text Input Field                | [Trash] |
+-----------------------------------------------------------------+
+ Validation message                                              |
+-----------------------------------------------------------------+
```

Both the `FormField` and `FormikFormField` accept a `layout` prop, and default to `StandardFieldLayout` if ommitted.

### CODE REVIEW

File: app/scripts/modules/core/src/presentation/forms/README.md [hunk 1]
```diff
@@ -1,130 +1,230 @@
...
-### Example Layout:
+### Form level validation Example:
 
-```
-+-----------------------------------------------------------------+
-+ * Label Text [Help] | Text Input Field                | [Trash] |
-+-----------------------------------------------------------------+
-+ Validation error message                                        |
-+-----------------------------------------------------------------+
+This example shows validation of a formik field using at the form level using the `Formik` component's `validate` prop.
```
Comment (jrsquared) [inline app/scripts/modules/core/src/presentation/forms/README.md]: `using at the form level using`? guessing remove `at the form level`?

File: app/scripts/modules/core/src/presentation/forms/README.md [hunk 2]
```diff
@@ -1,130 +1,230 @@
...
+      </Form>
+    );
+  }}
+/>
+```
 
-A layout component is composed of an Input and other field components.
-The layout defines how these components are rendered in relation to each other.
-The layout is also typically responsible for implementing any responsive behaviors.
+The `FormikFormField` component should be a decendent of a `Formik` component.
```
Comment (jrsquared) [inline app/scripts/modules/core/src/presentation/forms/README.md]: descendent

File: app/scripts/modules/core/src/presentation/forms/fields/FormikFormField.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+export class FormikFormField extends React.Component<IFormikFieldProps> {
+  public static defaultProps: Partial<IFormikFieldProps> = {
+    layout: StandardFieldLayout,
+  };
+
+  public render() {
+    const { input, layout, name, validate } = this.props; // ICommonFieldProps & name & validate
+    const { label, help, required, actions } = this.props; // IFieldLayoutPropsWithoutInput
```
Comment (jrsquared) [inline app/scripts/modules/core/src/presentation/forms/fields/FormikFormField.tsx]: i like the way you split this `this.props` destructuring.

File: app/scripts/modules/core/src/presentation/forms/fields/renderContent.tsx [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import * as React from 'react';
+
+export function renderContent<T>(Content: string | JSX.Element | React.ComponentType<T>, props: T) {
+  if (typeof Content === 'function' && Content.prototype && Content.prototype.isReactComponent) {
```
Comment (jrsquared) [inline app/scripts/modules/core/src/presentation/forms/fields/renderContent.tsx]: guessing you already checked, but does this catch `forwardRef` components?

File: app/scripts/modules/core/src/presentation/forms/inputs/TextInput.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+export class TextInput extends React.Component<ITextInputProps> {
+  public render() {
+    const { field, validation, inputClassName, ...otherProps } = this.props;
+    const fieldProps = { ...field, value: field.value || '' };
+
+    const validationClassName = classNames({
+      'form-control': true,
+      'ng-dirty': validation.touched,
```
Comment (jrsquared) [inline app/scripts/modules/core/src/presentation/forms/inputs/TextInput.tsx]: are there plans to move away from using `ng-*` classes?

File: app/scripts/modules/core/src/presentation/forms/CurrentForm.tsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ *   const username = formik.values.username;
+ *   return (
+ *     <h1>Hello {username}</h1>
+ *   )
+ * }} />
+ * ```
+ */
+export class CurrentForm<T = any> extends React.Component<ICurrentFormProps<T>> {
+  public render() {
+    return <Field name="" render={(props: FieldProps<T>) => this.props.render(props.form)} />;
```
Comment (jrsquared) [inline app/scripts/modules/core/src/presentation/forms/CurrentForm.tsx]: what does `name=""` do?

File: app/scripts/modules/core/src/presentation/forms/README.md [hunk 7]
```diff
@@ -1,130 +1,230 @@
...
+- accepts user input
+- emits change events when the input changes
+- visually indicates an invalid validation state
+
+An Input receives two important props (see `IFormInputProps`):
+
+- `field`: an `IControlledInputProps` object which provides convenient props for rendering controlled form inputs.
+  - `name`, `value`, `onChange`, `onBlur`
+  - The contents of this prop is typically spread onto an input, e.g., `<input {...field} />`
+- `validation`: The current validation state of the input used to visually indicate the current validation state of the input
```
Comment (jrsquared) [inline app/scripts/modules/core/src/presentation/forms/README.md]: department of redundancy department

File: app/scripts/modules/core/src/presentation/forms/fields/renderContent.tsx (hunk 4 above)
Comment (christopherthielen) [reply app/scripts/modules/core/src/presentation/forms/fields/renderContent.tsx]: no I didn't check.  I already forgot what forwardRef does.  will check

File: app/scripts/modules/core/src/presentation/forms/inputs/TextInput.tsx (hunk 5 above)
Comment (christopherthielen) [reply app/scripts/modules/core/src/presentation/forms/inputs/TextInput.tsx]: Yes, I plan to merge @gcomstock designs + this PR +styleguide in the near future.

File: app/scripts/modules/core/src/presentation/forms/CurrentForm.tsx (hunk 6 above)
Comment (christopherthielen) [reply app/scripts/modules/core/src/presentation/forms/CurrentForm.tsx]: I wanted a way to get access to the current `form` and used `<Field>` to do so because it's public API.  I think `name` is required prop to `<Field>`.  I'll check the docs and remove if it's not required.

File: app/scripts/modules/core/src/presentation/forms/README.md (hunk 7 above)
Comment (christopherthielen) [reply app/scripts/modules/core/src/presentation/forms/README.md]: This office will not tolerate redundancy in this office

File: app/scripts/modules/core/src/presentation/forms/CurrentForm.tsx (hunk 6 above)
Comment (christopherthielen) [reply app/scripts/modules/core/src/presentation/forms/CurrentForm.tsx]: Yes, it's required prop for `Field`.  It's basically a hack to get the `form` from a `Field` but not specify any actual field name.

File: app/scripts/modules/core/src/presentation/forms/fields/renderContent.tsx (hunk 4 above)
Comment (christopherthielen) [reply app/scripts/modules/core/src/presentation/forms/fields/renderContent.tsx]: not sure that this needs forwardRef support 🤔 
If we end up needing it we can add then
